scripts/cmake/config/default_ex.cmake

```diff
@@ -0,0 +1,129 @@
+# a default value. The goal is to encourage a "convention over configuration"
+# mindset for extension developers. If extension developers choose not follow
+# a suggested convention, they can override a configuration using two methods:
+#
+# 1) Specify each configuration as CMAKE_ARGS to the vmm_extension() macro. Ex:
+#
+#     vmm_extension(
+#           extension_name
+#           CMAKE_ARGS -DVMM_EX_BUILD_RULES=/path/to/build_rules.cmake
+#     )
+#
+# 2) Override a configuration in the extension's CMakeLists.txt file BEFORE
+#    calling include(${VMM_EXTENSION}). Ex:
+#
+#     set(VMM_EX_BUILD_RULES "/path/to/build_rules.cmake")
+#     set(VMM_EX_UNITTEST_PATH "/path/to/the/unit/tests")
+#     include(${VMM_EXTENSION})
+#
+
+# ------------------------------------------------------------------------------
+# Source tree structure
+# ------------------------------------------------------------------------------
+
+add_config(
+    CONFIG_NAME VMM_EX_TOP_DIR
+    CONFIG_TYPE PATH
+    DEFAULT_VAL ${CMAKE_CURRENT_SOURCE_DIR}
+    DESCRIPTION "Path to the top directory for this VMM extension"
+)
+
+add_config(
+    CONFIG_NAME VMM_EX_SOURCE_DIR
+    CONFIG_TYPE PATH
+    DEFAULT_VAL ${VMM_EX_TOP_DIR}/src
+    DESCRIPTION "Path to the source code directory for this VMM extension"
+    SKIP_VALIDATION
+)
+
+add_config(
+    CONFIG_NAME VMM_EX_INCLUDE_DIR
+    CONFIG_TYPE PATH
+    DEFAULT_VAL ${VMM_EX_TOP_DIR}/include
+    DESCRIPTION "Path to the header include directory for this VMM extension"
+    SKIP_VALIDATION
+)
+
+add_config(
+    CONFIG_NAME VMM_EX_UNITTEST_DIR
+    CONFIG_TYPE PATH
+    DEFAULT_VAL ${VMM_EX_TOP_DIR}/test
+    DESCRIPTION "Path to the unit test directory for this VMM extension"
+    SKIP_VALIDATION
+)
+
+# ------------------------------------------------------------------------------
+# Build system features
+# ------------------------------------------------------------------------------
+
+add_config(
+    CONFIG_NAME VMM_EX_CONFIGS
+    CONFIG_TYPE FILE
+    DEFAULT_VAL ${VMM_EX_TOP_DIR}/configs.cmake
+    DESCRIPTION "Path to extention-specific build configurations to be added "
+                "for this VMM extension"
+    SKIP_VALIDATION
+)
+
+add_config(
+    CONFIG_NAME VMM_EX_BUILD_RULES
+    CONFIG_TYPE FILE
+    DEFAULT_VAL ${VMM_EX_TOP_DIR}/build_rules.cmake
+    DESCRIPTION "Path to a build rules file to be validated before building "
+                "this VMM extension"
+    SKIP_VALIDATION
+)
+
+# ------------------------------------------------------------------------------
+# Read-only variables
+# ------------------------------------------------------------------------------
+
+add_config(
+    CONFIG_NAME VMM_EX_IS_UNITTEST_BUILD
+    CONFIG_TYPE INTERNAL
+    DEFAULT_VAL OFF
+    DESCRIPTION "The build system sets this variable to ON if this extension's "
+    "unit tests are currently being built, rather than it's source"
+    SKIP_VALIDATION
+)
+
+if(VMM_EX_IS_UNTTEST_BUILD)
```

Comment:
Is there a way to simplify this? Everything else is pretty nice to look at, and seems easy to follow, this one part stands out as "yuk".

Reply:
I agree that this is a bit "yuk", but I can't think of a simpler way around it. The real problem that this addresses is that there seems to be a pattern present in every VMM component (including the base components) that most of the source code needs to be re-built during a unit test build, but you still need to the ability to exclude a few things from a unit test build. With that in mind, I thought extension writters should have a variable available to check if the build system is currently building the extension normally, or if it was building the unit tests for the extension.

In the old build system, this same check was all over the place, but written as if(VMM_TOOLCHAIN_FILE AND ENABLE_UNITTESTING). I thought that was more yuk that having this variable available. 

Any alternative suggestions?

Reply:
Your fine here as well since this is not something the user will be coding themselves.